internal/fsck/fsck.go

```diff
@@ -0,0 +1,241 @@
+// Copyright 2025 The Tessera authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fsck
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/transparency-dev/merkle/compact"
+	"github.com/transparency-dev/merkle/rfc6962"
+	"github.com/transparency-dev/trillian-tessera/api"
+	"github.com/transparency-dev/trillian-tessera/api/layout"
+	"github.com/transparency-dev/trillian-tessera/internal/parse"
+	"github.com/transparency-dev/trillian-tessera/internal/stream"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/klog/v2"
+)
+
+// Fetcher describes a struct which knows how to retrieve tlog-tiles artifacts from a log.
+type Fetcher interface {
+	ReadCheckpoint(ctx context.Context) ([]byte, error)
+	ReadTile(ctx context.Context, l, i uint64, p uint8) ([]byte, error)
+	ReadEntryBundle(ctx context.Context, i uint64, p uint8) ([]byte, error)
+}
+
+// Check performs an integrity check against a log via the provided fetcher, using the provided
+// bundleHasher to parse and convert entries from the log's entry bundles into leaf hashes.
+//
+// The leaf hashes are used to:
+// 1. re-construct the root hash of the log, and compare it against the value in the log's checkpoint
+// 2. re-construct the internal tiles of the log, and compare them against the log's tile resources.
+//
+// The checking will use the provided N parameter to control the number of concurrent workers undertaking
+// this process.
+func Check(ctx context.Context, f Fetcher, N uint, bundleHasher func([]byte) ([][]byte, error)) error {
+	cp, err := f.ReadCheckpoint(ctx)
+	if err != nil {
+		klog.Exitf("fetch initial source checkpoint: %v", err)
+	}
+	klog.V(1).Infof("Fsck: checkpoint:\n%s", cp)
+	// TODO(al): Should really open this with the pubK to verify the checkpoint is well formed.
```

Comment:
IMO it's worth doing this. The more we do this unsafe parsing, the more it will become a dangerous norm.